netcreateorg/netcreate-2018
# Node/Edge Option Type Editor

Merge after #198!
---
Branch: `dev-bl/typeedit`

This hacks the json-editor interface so that we can add some additional functionality around:
* Renaming an existing node/edge type option label
* Deleting a type option label
* Deleting a type option label and remapping the deleted label to another label -- you would use this to combine two labels into one.

In addition to updating the template, the editor will now also update the existing dataset's type options if necessary.

---

# To Test

We recommend making a backup of your dataset and template because these tests will change both template and dataset content.

1. `git fetch && git checkout dev-bl/typeedit`
2. `npm run dev`
3. Go to `http://localhost:3000`

### Test Rename

To rename a label, you just type the new name in the "Change To" field.  The nice thing about this UI is that the original label is still visible if you're making decisions about naming while reviewing the whole list.

4. Select "More..." and "Edit Template"
5. Click on "Edit Node Types"
6. Rename "Person" to "People"
7. Click "Save Changes"
8. Any nodes marked as Person should still be the same color (light blue by default)
9. Open a known Person node and make sure its type has been changed to "People"
10. In the terminal, `ctrl-c` quit NetCreate and start it again.
11. The fresh reload of the data should still show the changes: Node types have changed to "People" and the color remains the same.



### Test Delete

You can simply remove a type option by deleting it.  However, deleting a type option will make it an unknown type -- this means that any node that is set to that type will be displayed as black because it will have an unknown type.  We expect that most of the time when you want to delete a label, you would also want to remap it (see the next section "Test Delete and Remap").

12. Select "More..." and "Edit Template"
13. Click on "Edit Node Types"
14. Check the "Delete" next to "Place" or some other type that you have on screen.
15. Click "Save Changes"
16. Any node that was a "Place" node should now display as black, since "Place" is no longer a type the system recognizes.

### Test Delete and Remap

Instead of simply deleting a type option, you can tell the system to change any node with the deleted type to a new type.

12. Select "More..." and "Edit Template"
13. Click on "Edit Node Types"
14. Check the "Delete" next to "Group" or some other type that you have on screen.
15. In the "Change To" field, enter an existing name to map to, e.g. "People"
16. Click "Save Changes"
17. Any node that was a "Group" node should now display as a blue "People" node.  The "Group" type is now deleted.

### Test Adding

You can add new options.

12. Select "More..." and "Edit Template"
13. Click on "Edit Node Types"
14. Check the "Add row".
15. In the "Label" field, enter "Race"
16. Leave the "Change To" field blank.
17. Select a color
18. Click "Save Changes"
19. Set a node to the new type.
20. You should see the new type listed in the type menu, and once it's set you should see the new color.

Technically we should probably hide the "Change To" field when adding a new field, but having it actually adds another handy feature.  Read on...

### Remap Inadvertent "Feature"

Another handy power user feature for managing type options is you can use the "Add Row" feature to both add a type AND rename it at the same time.  This is handy if you find that you have a node with a stray type that hasn't been added.  For example, let's say you found three black nodes set to a type "Race", but you do not currently have "Race" as a type option.  And you don't want the type option to be "Race", you want it to be "Ethnicity".  You can Add it and change it in one action:

12. Select "More..." and "Edit Template"
13. Click on "Edit Node Types"
14. Check the "Add row".
15. In the "Label" field, enter "Race"
16. In the "Change To" field, enter "Ethnicity"
17. Set the color to Purple.
18. Click "Save Changes"
17. Any node that was a "Race" node should now display as a purple "Ethnicity" node.  

Under the hood what happens is:
1. We create a new type option for any "Label" we find in the list.
2. Then if there is something in the "Change To" field, we will search for nodes with the "Label" type and change them to the "Change To" type.

The UI is a little quirky, especially for someone who is not used to managing lists of items in this way.  But the nice thing is we don't need to build a custom UI with a ton of actions to take for every single label.

---


# Wiki

# Edit Template

## Locking Templates

While editing templates, you cannot edit nodes or edges.  This is because changes to the Template can affect how Nodes and Edges behave and are displayed.  Changes to the template can also result in changes to the dataset (e.g. one node type is remapped to another node type).

Conversely, when you are editing a node or edge, you are not allowed to edit templates.

There is a warning message displayed on the Node/Edge if this occurs.


## Edit Node Types / Edit Edge Types

When editing node types or edge types, if you use the "Change To" field, Net.Create will automatically update your dataset for you: any nodes and edges using the original "Label" type will be updated to use the new "Change To" type.  This change is executed as soon as you click "Save Changes".  Your dataset file on disk will be updated immediately.


### NOTES

* If you inadvertently add a row and leave the "Label" field blank, the system will remove the row when it saves.
* The system automatically adds a "blank default" type option so that Node Types and Edge Types can be set to an un-selected blank state.  This is completely invisible to the user.
